In [56]:
import pandas as pd
import numpy as np
import math

# Read in the data
df = pd.read_csv('/Users/cl3720/Desktop/copilot-examples/GPT_Gene_Prioritization/Data/Results/GPT4_top_prediction.csv')
# split columns into multiple columns
df['dataset'] = df['file_path'].str.split('/', expand=True)[8]

In [33]:
# format float to 2 decimal places
w = df.shape[0]
x = df[['predict_correct_in_top_5']].values.sum()
y = df[['predict_made_in_top_10']].values.sum()
print('{y} out of {w} predictions made in top 5 ({z}%)'.format(y=y, w=w, z=np.round(y/w,4)*100))
print('{x} out of {y} predictions correct in top 5 ({z}%)'.format(x=x, y=y, z=np.round(x/y,4)*100))

x = df[['predict_correct_in_top_10']].values.sum()
y = df[['predict_made_in_top_10']].values.sum()
print('{y} out of {w} predictions made in top 10 ({z}%)'.format(y=y, w=w, z=np.round(y/w,4)*100))
print('{x} out of {y} predictions correct in top 10 ({z}%)'.format(x=x, y=y, z=np.round(x/y,4)*100))

x = df[['predict_correct_in_top_50']].values.sum()
y = df[['predict_made_in_top_50']].values.sum()
print('{y} out of {w} predictions made in top 50 ({z}%)'.format(y=y, w=w, z=np.round(y/w,4)*100))
print('{x} out of {y} predictions correct in top 50 ({z}%)'.format(x=x, y=y, z=np.round(x/y,4)*100))


253 out of 276 predictions made in top 5 (91.67%)
28 out of 253 predictions correct in top 5 (11.07%)
253 out of 276 predictions made in top 10 (91.67%)
34 out of 253 predictions correct in top 10 (13.44%)
195 out of 276 predictions made in top 50 (70.65%)
21 out of 195 predictions correct in top 50 (10.77%)


In [41]:
# subset df by dataset
dataset_list = df['dataset'].unique()
for dataset in dataset_list:
    print(dataset)
    df_subset = df[df['dataset']==dataset]
    w = df_subset.shape[0]
    x = df_subset[['predict_correct_in_top_5']].values.sum()
    y = df_subset[['predict_made_in_top_5']].values.sum()
    print('{y} out of {w} predictions made in top 5 ({z}%)'.format(y=y, w=w, z=np.round(y/w,4)*100))
    print('{x} out of {y} predictions correct in top 5 ({z}%)'.format(x=x, y=y, z=np.round(x/y,4)*100))

    x = df_subset[['predict_correct_in_top_10']].values.sum()
    y = df_subset[['predict_made_in_top_10']].values.sum()
    print('{y} out of {w} predictions made in top 10 ({z}%)'.format(y=y, w=w, z=np.round(y/w,4)*100))
    print('{x} out of {y} predictions correct in top 10 ({z}%)'.format(x=x, y=y, z=np.round(x/y,4)*100))

    x = df_subset[['predict_correct_in_top_50']].values.sum()
    y = df_subset[['predict_made_in_top_50']].values.sum()
    print('{y} out of {w} predictions made in top 50 ({z}%)'.format(y=y, w=w, z=np.round(y/w,4)*100))
    print('{x} out of {y} predictions correct in top 50 ({z}%)'.format(x=x, y=y, z=np.round(x/y,4)*100))
    print('')


ColumbiaU
27 out of 27 predictions made in top 5 (100.0%)
6 out of 27 predictions correct in top 5 (22.220000000000002%)
27 out of 27 predictions made in top 10 (100.0%)
7 out of 27 predictions correct in top 10 (25.929999999999996%)
22 out of 27 predictions made in top 50 (81.47999999999999%)
4 out of 22 predictions correct in top 50 (18.18%)

CSH
70 out of 72 predictions made in top 5 (97.22%)
12 out of 70 predictions correct in top 5 (17.14%)
63 out of 72 predictions made in top 10 (87.5%)
13 out of 63 predictions correct in top 10 (20.630000000000003%)
49 out of 72 predictions made in top 50 (68.06%)
10 out of 49 predictions correct in top 50 (20.41%)

AJHG
74 out of 78 predictions made in top 5 (94.87%)
0 out of 74 predictions correct in top 5 (0.0%)
71 out of 78 predictions made in top 10 (91.03%)
0 out of 71 predictions correct in top 10 (0.0%)
51 out of 78 predictions made in top 50 (65.38000000000001%)
1 out of 51 predictions correct in top 50 (1.96%)

DGD
82 out of 85 predict

In [51]:
def probability_of_picking_specific_card(total_cards, specific_cards, draws):
    probability_not_picking = 1

    for i in range(draws):
        non_specific_cards = total_cards - specific_cards - i
        remaining_cards = total_cards - i
        probability_not_picking *= non_specific_cards / remaining_cards

    probability_picking = 1 - probability_not_picking
    return probability_picking
# calculate the probability of select the correct gene in top 5

total_cards = 30000
specific_cards = 1
# what is the probability of selecting the correct gene in top 5
draws = 5
probability = probability_of_picking_specific_card(total_cards, specific_cards, draws)
print('Probability of selecting the correct gene in top 5: {}%'.format(probability*100))
draws = 10
probability = probability_of_picking_specific_card(total_cards, specific_cards, draws)
print('Probability of selecting the correct gene in top 10: {}%'.format(probability*100))
draws = 50
probability = probability_of_picking_specific_card(total_cards, specific_cards, draws)
print('Probability of selecting the correct gene in top 50: {}%'.format(probability*100))

Probability of selecting the correct gene in top 5: 0.01666666666666483%
Probability of selecting the correct gene in top 10: 0.033333333333340764%
Probability of selecting the correct gene in top 50: 0.16666666666667052%


In [52]:
# This table is generated using ChatGPT-3

| Dataset   | Top 5          |                | Top 10         |                | Top 50          |                |
|-----------|----------------|-------------- |----------------|-------------- |-----------------|-------------- |
|           | Predictions (%)| Correct (%)   | Predictions (%)| Correct (%)   | Predictions (%)| Correct (%)   |
| ColumbiaU | 27 (100.00%)   | 6 (22.22%)    | 27 (100.00%)   | 7 (25.93%)    | 22 (81.48%)     | 4 (18.18%)    |
| CSH       | 70 (97.22%)    | 12 (17.14%)   | 63 (87.50%)    | 13 (20.63%)   | 49 (68.06%)     | 10 (20.41%)   |
| AJHG      | 74 (94.87%)    | 0 (0.00%)     | 71 (91.03%)    | 0 (0.00%)     | 51 (65.38%)     | 1 (1.96%)     |
| DGD       | 82 (96.47%)    | 10 (12.20%)   | 81 (95.29%)    | 14 (17.28%)   | 62 (72.94%)     | 6 (9.68%)     |
| TAF1      | 8 (57.14%)     | 0 (0.00%)     | 11 (78.57%)    | 0 (0.00%)     | 11 (78.57%)     | 0 (0.00%)     |


In [64]:
# summary the prediction results by gene
df.groupby('dx_gene').agg({'predict_correct_in_top_5': ['sum', 'count', lambda x: sum(x)/len(x)], 'predict_correct_in_top_10': ['sum', 'count'], 'predict_correct_in_top_50': ['sum', 'count']}).sort_values(('predict_correct_in_top_5', 'count'), ascending=False).head(20)

predict_correct_in_top_5                  predict_correct_in_top_10   
                             sum count <lambda_0>                       sum   
dx_gene                                                                       
TAF1                           0    14       0.00                         0  \
WDR26                          0    13       0.00                         0   
DHX30                          0    12       0.00                         0   
TP53                           8     8       1.00                         8   
CDK10                          0     8       0.00                         0   
YWHAG                          0     6       0.00                         0   
MRPS34                         0     6       0.00                         0   
FDXR                           0     6       0.00                         0   
RAB11B                         0     5       0.00                         0   
ARID1B                         1     4       0.25                         1   
GRM1                           0     4       0.00                         0   
RAC1                           0     4       0.00                         0   
SYNGAP1                        1     3       0.33                         1   
PURA                           0     3       0.00                         0   
CDKL5                          2     3       0.67                         2   
BRAF                           0     3       0.00                         0   
SPART                          0     3       0.00                         0   
LIPT2                          0     3       0.00                         0   
TBC1D23                        0     3       0.00                         0   
NPR3                           0     3       0.00                         0   

              predict_correct_in_top_50        
        count                       sum count  
dx_gene                                        
TAF1       14                         0    14  
WDR26      13                         0    13  
DHX30      12                         0    12  
TP53        8                         5     8  
CDK10       8                         0     8  
YWHAG       6                         0     6  
MRPS34      6                         0     6  
FDXR        6                         0     6  
RAB11B      5                         0     5  
ARID1B      4                         0     4  
GRM1        4                         1     4  
RAC1        4                         0     4  
SYNGAP1     3                         1     3  
PURA        3                         0     3  
CDKL5       3                         1     3  
BRAF        3                         0     3  
SPART       3                         0     3  
LIPT2       3                         0     3  
TBC1D23     3                         0     3  
NPR3        3                         0     3

In [65]:
df.groupby('dx_gene').agg({'predict_correct_in_top_5': ['sum', 'count', lambda x: sum(x)/len(x)], 'predict_correct_in_top_10': ['sum', 'count'], 'predict_correct_in_top_50': ['sum', 'count']}).sort_values(('predict_correct_in_top_5', 'sum'), ascending=False).head(20)

predict_correct_in_top_5                  predict_correct_in_top_10   
                             sum count <lambda_0>                       sum   
dx_gene                                                                       
TP53                           8     8       1.00                         8  \
SCN1A                          2     2       1.00                         2   
CDKL5                          2     3       0.67                         2   
ARID1B                         1     4       0.25                         1   
SHH                            1     1       1.00                         1   
MMUT                           1     1       1.00                         1   
COL18A1                        1     1       1.00                         0   
PTEN                           1     2       0.50                         1   
SOX9                           1     1       1.00                         1   
SPG11                          1     1       1.00                         1   
SPTA1                          1     1       1.00                         1   
SYNGAP1                        1     3       0.33                         1   
RYR1                           1     1       1.00                         1   
PAX6                           1     1       1.00                         1   
MECP2                          1     1       1.00                         1   
ACTG2                          1     1       1.00                         1   
ANK1                           1     1       1.00                         1   
WFS1                           1     1       1.00                         1   
MYO7A                          1     1       1.00                         1   
NSD1                           0     1       0.00                         0   

              predict_correct_in_top_50        
        count                       sum count  
dx_gene                                        
TP53        8                         5     8  
SCN1A       2                         1     2  
CDKL5       3                         1     3  
ARID1B      4                         0     4  
SHH         1                         1     1  
MMUT        1                         1     1  
COL18A1     1                         0     1  
PTEN        2                         1     2  
SOX9        1                         0     1  
SPG11       1                         1     1  
SPTA1       1                         0     1  
SYNGAP1     3                         1     3  
RYR1        1                         0     1  
PAX6        1                         0     1  
MECP2       1                         0     1  
ACTG2       1                         0     1  
ANK1        1                         0     1  
WFS1        1                         0     1  
MYO7A       1                         0     1  
NSD1        1                         0     1

In [ ]:
# gene, predicted_positive, true_positive, precision, recall, f1